src/net.cpp

```diff
@@ -902,26 +902,44 @@ void ProtectEvictionCandidatesByRatio(std::vector<NodeEvictionCandidate>& vEvict
     size_t total_protect_size = initial_size / 2;
-    const size_t onion_protect_size = total_protect_size / 2;
 
-    if (onion_protect_size) {
-        // Pick out up to 1/4 peers connected via our onion service, sorted by longest uptime.
-        EraseLastKElements(vEvictionCandidates, CompareOnionTimeConnected, onion_protect_size,
-                           [](const NodeEvictionCandidate& n) { return n.m_is_onion; });
+    // Count number of nodes connected via localhost and onion.
+    // Higher positions in the array recover unused slots left by lower ones.
+    std::array<size_t, 2> network_counts{{}};
+    for (const NodeEvictionCandidate& node : vEvictionCandidates) {
+        if (node.m_is_local) {
+            ++network_counts[0];
+        } else if (node.m_is_onion) {
+            ++network_counts[1];
+        }
     }
 
-    const size_t localhost_min_protect_size{2};
-    if (onion_protect_size >= localhost_min_protect_size) {
-        // Allocate any remaining slots of the 1/4, or minimum 2 additional slots,
-        // to localhost peers, sorted by longest uptime, as manually configured
-        // hidden services not using `-bind=addr[:port]=onion` will not be detected
-        // as inbound onion connections.
-        const size_t remaining_tor_slots{onion_protect_size - (initial_size - vEvictionCandidates.size())};
-        const size_t localhost_protect_size{std::max(remaining_tor_slots, localhost_min_protect_size)};
-        EraseLastKElements(vEvictionCandidates, CompareLocalHostTimeConnected, localhost_protect_size,
-                           [](const NodeEvictionCandidate& n) { return n.m_is_local; });
-    }
+    // Find the number of networks from which we have disadvantaged candidates.
+    const auto networks = std::count_if(network_counts.begin(), network_counts.end(), [](auto n) { return n; });
+
+    if (networks) {
+        // Obtain the minimum number of peers per network to potentially protect.
+        // The base ratio is 1/2 (for the case of peers from only 1 network).
+        const size_t protect_size_per_network{total_protect_size / (2 * networks)};
 
+        if (protect_size_per_network) {
+            size_t total_to_protect{protect_size_per_network};
+
+            if (network_counts[0]) {
+                // Protect localhost peers, sorted by longest uptime.
+                EraseLastKElements(vEvictionCandidates, CompareLocalHostTimeConnected, total_to_protect,
+                                   [](const NodeEvictionCandidate& n) { return n.m_is_local; });
+                total_to_protect += protect_size_per_network;
+            }
+
+            if (network_counts[1]) {
+                // Protect onion peers (connected via our onion service), sorted by longest uptime.
+                const size_t onion_peers_to_protect{total_to_protect - (initial_size - vEvictionCandidates.size())};
+                EraseLastKElements(vEvictionCandidates, CompareOnionTimeConnected, onion_peers_to_protect,
```

Comment:
Since the inital number of protected slots per network goes down quickly as the number of networks increases, it seems good to not waste any unused slots in this basic greedy knapsack. So we calculate `peers_to_protect` for each iteration having potential peers to protect, in order to recover any unused slots from the previous iteration.

Reply:
Hm, `s/total_to_protect/available_to_protect/`?